File: typescript/src/croissant.ts
export interface FieldInit { name: string; dataType: string; description: string; semanticTypeValue?: string }
export class Field {
  readonly name: string; readonly dataType: string; readonly description: string; readonly semanticTypeValue?: string;
  constructor(init: FieldInit) { this.name = init.name; this.dataType = init.dataType; this.description = init.description; this.semanticTypeValue = init.semanticTypeValue; }
  semanticType(value: string): Field { return new Field({ ...this, semanticTypeValue: value }); }
}

export interface FileObjectInit { id: string; name: string; contentUrl: string; encodingFormat: string }
export class FileObject { readonly id: string; readonly name: string; readonly contentUrl: string; readonly encodingFormat: string;
  constructor(init: FileObjectInit) { Object.assign(this, init); }
}

export interface RecordSetInit { id: string; name: string; fields?: Field[] }
export class RecordSet { readonly id: string; readonly name: string; readonly fields: Field[];
  constructor(init: RecordSetInit) { this.id = init.id; this.name = init.name; this.fields = init.fields ?? []; }
}

export interface CroissantDatasetInit { id: string; name: string; description: string; license: string; creators: string[]; files: FileObject[]; recordSets: RecordSet[]; keywords: string[] }
export class CroissantDataset {
  readonly id: string; readonly name: string; readonly description: string; readonly license: string; readonly creators: string[]; readonly files: FileObject[]; readonly recordSets: RecordSet[]; readonly keywords: string[];
  constructor(init: CroissantDatasetInit) { Object.assign(this, init); }
  toJsonLd(): Record<string, unknown> {
    return { "@context": { "@vocab": "https://schema.org/", cr: "http://mlcommons.org/croissant/", dcat: "http://www.w3.org/ns/dcat#", odrl: "http://www.w3.org/ns/odrl/2/" }, "@type": "cr:Dataset", "@id": this.id, name: this.name, description: this.description, license: this.license, creator: this.creators.map((name) => ({ "@type": "Person", name })), keywords: this.keywords,
      distribution: this.files.map((file) => ({ "@type": "cr:FileObject", "@id": file.id, name: file.name, contentUrl: file.contentUrl, encodingFormat: file.encodingFormat })),
      recordSet: this.recordSets.map((recordSet) => ({ "@type": "cr:RecordSet", "@id": recordSet.id, name: recordSet.name, field: recordSet.fields.map((field) => ({ "@type": "cr:Field", name: field.name, dataType: field.dataType, description: field.description, sameAs: field.semanticTypeValue ?? null })) })),
    };
  }
}

File: src/croissant.rs
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Field {
    pub name: String,
    pub data_type: String,
    pub description: String,
    pub semantic_type: Option<String>,
}

impl Field {
    pub fn new(
        name: impl Into<String>,
        data_type: impl Into<String>,
        description: impl Into<String>,
    ) -> Self {
        Self {
            name: name.into(),
            data_type: data_type.into(),
            description: description.into(),
            semantic_type: None,
        }
    }

    pub fn semantic_type(mut self, semantic_type: impl Into<String>) -> Self {
        self.semantic_type = Some(semantic_type.into());
        self
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct FileObject {
    pub id: String,
    pub name: String,
    pub content_url: String,
    pub encoding_format: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordSet {
    pub id: String,
    pub name: String,
    pub fields: Vec<Field>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct CroissantDataset {
    pub id: String,
    pub name: String,
    pub description: String,
    pub license: String,
    pub creators: Vec<String>,
    pub files: Vec<FileObject>,
    pub record_sets: Vec<RecordSet>,
    pub keywords: Vec<String>,
}

impl CroissantDataset {
    pub fn to_json_ld(&self) -> Value {
        json!({
            "@context": {
                "@vocab": "https://schema.org/",
                "cr": "http://mlcommons.org/croissant/",
                "dcat": "http://www.w3.org/ns/dcat#",
                "odrl": "http://www.w3.org/ns/odrl/2/"
            },
            "@type": "cr:Dataset",
            "@id": self.id,
            "name": self.name,
            "description": self.description,
            "license": self.license,
            "creator": self.creators.iter().map(|name| json!({"@type": "Person", "name": name})).collect::<Vec<_>>(),
            "keywords": self.keywords,
            "distribution": self.files.iter().map(|file| {
                json!({
                    "@type": "cr:FileObject",
                    "@id": file.id,
                    "name": file.name,
                    "contentUrl": file.content_url,
                    "encodingFormat": file.encoding_format
                })
            }).collect::<Vec<_>>(),
            "recordSet": self.record_sets.iter().map(|record_set| {
                json!({
                    "@type": "cr:RecordSet",
                    "@id": record_set.id,
                    "name": record_set.name,
                    "field": record_set.fields.iter().map(|field| {
                        json!({
                            "@type": "cr:Field",
                            "name": field.name,
                            "dataType": field.data_type,
                            "description": field.description,
                            "sameAs": field.semantic_type
                        })
                    }).collect::<Vec<_>>()
                })
            }).collect::<Vec<_>>()
        })
    }
}
